### PR TITLE
feat: use derived username instead of email for non-member superadmins

### PR DESCRIPTION
**Pull request by rubenfiszel**
## Summary

Fixes WIN-2115.

When a superadmin accesses a workspace they are **not** a member of, Windmill previously used their raw **email** as their username when constructing `ApiAuthed`/`Authed`, leaking the email downstream (`created_by`, `permissioned_as`, folder owner, audit actor, draft author, etc.).

Almost every instance has `automate_username_creation` enabled, which derives and assigns a stable instance-wide username (`password.username`). This PR uses that derived username instead of the email for non-member superadmins.

**Fully conditional on the `automate_username_creation` instance setting.** When disabled (`password.username` is `NULL`) behaviour is byte-for-byte unchanged (verified). **Only superadmin non-members are affected** — regular members and superadmins who *are* members are untouched (verified).

## Changes

- **`windmill-common/src/usernames.rs`**: `get_instance_username_or_fallback_to_email` — derived username when set, else email.
- **`windmill-api-auth/src/auth.rs`** + **`windmill-common/src/auth.rs`**: the two auth sites that build a non-member superadmin's identity now use the derived username (covers token auth, JWT-mint, and job-perms paths).
- **`windmill-common/src/users.rs`**: new shared `resolve_username_to_email(w_id, username, db)` — resolves via `usr`, then falls back to `password` (matching both the derived username *and* the email, gated on `super_admin`). Used by `get_email_from_permissioned_as` so schedules/triggers/folders resolve a non-member superadmin's real email (was returning a bogus `<username>@unknown.windmill.dev`).
- **`windmill-api/src/drafts.rs`**: `get_draft_for_user` and the `list_drafts` author list now use the resolver / a `password` fallback instead of an `admins`-workspace hack that assumed *username == email*. Fixes drafts 404-ing for non-member superadmins and stops the raw email leaking into the drafts payload.
- **`windmill-api-users/src/users.rs`**: `whoami` returns the derived username, guards membership by email match, and adds an explicit `non_member: bool`.
- **`openapi.yaml` + frontend**: expose `non_member`; `UserMenu.svelte` shows "Superadmin, not a member of this workspace" and "You act as `<username>` here".

## Safety analysis — dogfooded end-to-end

Jobs store the acting email in a dedicated `permissioned_as_email` column, so the real email is always preserved; only the fields that are *supposed* to be usernames change.

**Only superadmin non-members are affected:**

| identity | `whoami` | job `created_by` / `permissioned_as` | resolver fallback |
|---|---|---|---|
| regular member (alice) | `non_member=false`, `alice` | `alice` / `u/alice` | via `usr` (no fallback) |
| superadmin **member** | `non_member=false`, real role | via `usr` | via `usr` |
| superadmin **non-member** (Mode A) | `non_member=true`, `admin` | `admin` / `u/admin` | `password` → real email |
| non-superadmin non-member | — | — | **not resolved** (gate blocks) |

**Mode A vs Mode B** for the non-member superadmin:

| field | Mode A (enabled) | Mode B (disabled) |
|---|---|---|
| `whoami.username` | `admin` | `admin@windmill.dev` |
| `created_by` / `permissioned_as` | `admin` / `u/admin` | `admin@windmill.dev` / `admin@windmill.dev` |
| `permissioned_as_email` | `admin@windmill.dev` | `admin@windmill.dev` |
| `get_email_from_permissioned_as` | `u/admin` → real email | raw email (unchanged) |
| drafts `get`/`list` | resolves; author shown as `admin` | resolves; author shown as email |

Mode B is identical to pre-PR behaviour. Other username→email lookups (`username_to_email` endpoint, offboarding, user-management) operate on explicit members or were already 404 for non-member superadmins — not regressions.

## Notes / decisions

- **`non_member` boolean vs `role` string**: an explicit boolean is unambiguous (`usr.role` is free-text) and lives correctly on `UserInfo` (always `false` for member rows).
- **Caching**: no new `email->username` cache — hot auth paths already have 120s caches; `get_email_from_permissioned_as` keeps its existing `EMAIL_CACHE`.
- **Superadmin gate on the reverse resolver**: mirrors the forward direction exactly (only superadmins produce non-member `u/{username}` values), so a removed regular member's stale `permissioned_as` still resolves to `@unknown` as before rather than changing behaviour.

## Screenshots

usermenu-superadmin-nonmember

## Test plan

- [x] Superadmin non-member (Mode A): job stores `created_by=admin`, `permissioned_as=u/admin`, email preserved; `get_email_from_permissioned_as('u/admin')` → real email; drafts `get`/`list` resolve and show `admin` (no email leak).
- [x] Regular member (alice) and superadmin-as-member: `non_member=false`, all identity via `usr` — unaffected.
- [x] Resolver gate: non-superadmin non-member does **not** resolve via the `password` fallback.
- [x] Mode B (setting disabled): every surface identical to pre-PR (email everywhere).
- [x] Frontend submenu clarification; sidebar shows `User (admin)`.
- [x] sqlx offline cache regenerated (net +1 query; EE caches preserved).

## Test suite

- Updated `test_schedule_permissions_superadmin_not_in_workspace` (in `preserve_on_behalf_of.rs`): it asserted the *old* behaviour (email as `permissioned_as` for a superadmin not in the workspace); now asserts `permissioned_as = u/<derived>` with the email still stored on the schedule. Passing.
- Verified passing: `preserve_on_behalf_of` (9), `trigger_listener_queries` (4), `folder_default_permissioned_as`, `jobs_read_auth`, `drafts_nul`, and `windmill-common`/`windmill-api-users` unit tests.
- `asset_trigger_dispatch` (7 failures) is a **pre-existing** flake unrelated to this change — the identical 7 tests fail on `origin/main` with these changes reverted (MIN_VERSION stays `0.0.0` in test binaries; no username/email/permissioned_as involvement).

## Exhaustive audit (username↔email surfaces)

Swept every backend query touching `usr` by `username`, every username→email resolver, every email→username display join, and every superadmin-identity construction site (two independent agent passes + manual grep). Found and fixed:

- **Draft owner-circles** in `scripts.rs`, `flows.rs`, `apps.rs`, and `user_drafts.rs` (`fetch_other_drafts_users`) — same `CASE WHEN workspace_id='admins' THEN email` hack as the drafts endpoints; now use the `COALESCE(u.username, p.username, …)` password fallback. The `admins`-hack pattern now has **zero** occurrences in the tree.
- **`username_to_email` endpoint** — was a raw `usr` lookup (404 for a non-member superadmin's derived username); now uses `resolve_username_to_email`.
- **Home items-list filter** (`ItemsList.svelte`) — used `is_super_admin && username.includes('@')` as a proxy for "non-member superadmin", broken in Mode A; now uses the `non_member` flag.
- **Favorites backfill migration** — a non-member superadmin's pre-upgrade favorites were keyed by email; a scoped, guarded `UPDATE` realigns them to the derived username so they stay visible.

Verified SAFE (no change needed): job/schedule/trigger/app/SMTP push (all route through the resolver); the two `u/{username}` auth branches (guarded by `super_admin`); offboarding / update / delete / get workspace-user (member-directory, and non-regressions — they 404'd on the email before too); folder-rule validation (member-only by design); JWT / superadmin_secret / agent-worker / MCP identity (synthetic or inherit the fixed `ApiAuthed`); login/logout/token audit authors (email-as-actor is intentional, pre-session); all `username.split('@')`/`contains('@')` path-prefix helpers (produce the same result and are still needed for the setting-disabled mode).

### CODE REVIEW

File: backend/windmill-common/src/usernames.rs [hunk 1]
```diff
@@ -94,6 +94,34 @@ pub async fn generate_instance_username_for_all_users(db: &DB) -> error::Result<
...
+{
+    match sqlx::query_scalar!("SELECT username FROM password WHERE email = $1", email)
+        .fetch_optional(db)
+        .await
+    {
+        Ok(Some(Some(username))) => username,
+        // No derived username (automate_username_creation off), or a transient DB
+        // error: fall back to the email for this call. On error we intentionally
+        // do not persist anything, so the next call retries.
+        _ => email.to_string(),
```
[the remark this answers is not in the document]
Comment (rubenfiszel) [reply backend/windmill-common/src/usernames.rs]: Fixed in 9249b043f9 — `get_instance_username_or_fallback_to_email` now returns `Result<String>` and only falls back to the email for a genuine "no derived username" (`Ok(None)`); a query error propagates so both callers fail closed (the token path returns `None`/401, the `fetch_authed` path returns `Err`) rather than leaking the raw email as the acting username.

**Comment by rubenfiszel** [comment]
Re: the Codex Review above — I investigated both P1s and neither is a valid issue for this PR.

**1 — preserved `permissioned_as` resolving to a superadmin.** Pre-existing and unchanged by this PR. Supplying an arbitrary `permissioned_as` is already gated to admins / `wm_deployers` (`can_preserve_on_behalf_of`), and `resolve_permissioned_as` stores it verbatim with no membership check. On `main` today the **bare-email** spelling (`permissioned_as = "superadmin@windmill.dev"`) already resolves through the `fetch_authed_from_permissioned_as` else-branch to `is_admin = is_super_admin = true` — so the same privileged actors could already produce a superadmin-owned job before this change. The password fallback only adds an equivalent `u/<derived>` spelling for an identity those actors could already target via the email form. No new actor, boundary, or capability is introduced. Constraining the fallback would not close the pre-existing bare-email path, and would break resolving a superadmin's *own* legitimate stored identity (their schedules/jobs/drafts).

**2 — stale `password.username` in disabled mode.** The collision cannot occur in practice. While `automate_username_creation` is enabled, usernames are globally unique — `generate_instance_wide_unique_username` checks `usr` (across all workspaces) ∪ `password` ∪ `pending_user` — so no member can ever be assigned a name colliding with a superadmin's `password.username`. The setting is not turned back off once enabled (it defaults on; the pre-write hook only backfills on enable and never clears on disable), so the disabled-after-enabled regime the finding depends on is never entered.

Given both edges are pre-existing / unreachable, the two suggested edge-case tests aren't warranted. (Pi's review reached the same "good to merge" conclusion independently.)

**Comment by rubenfiszel** [comment]
Both latest Codex P1s are now addressed in code:

**1 — `username_to_email` disclosing a non-member superadmin's email.** Valid — I'd routed that endpoint through the superadmin `password` fallback in an earlier round, but it has no target/superadmin gate, so any workspace-authenticated caller could turn a guessed derived username into a non-member superadmin's email. Reverted the endpoint to the member-only `usr` lookup (`fa93295ed0`); it 404s again for non-members. Internal schedule/trigger/draft resolution keeps the fallback via `resolve_username_to_email` and never returns the email to an arbitrary caller.

**2 — stale `password.username` after disabling the setting.** Fair point that our earlier "can't be disabled" argument was a UI convention, not enforced. Made it real (`10418f2ffb`): `automate_username_creation` can no longer be set to `false` once any instance-wide username has been assigned — so the global-uniqueness guarantee that keeps stored `u/<username>` identities unambiguous can't be dropped to workspace-local uniqueness. Re-saving `false` on an already-disabled instance stays a no-op, so disabled-mode instances aren't broken. Verified: disable-while-enabled → 400 with a clear message; re-save-while-disabled → 200.

**Comment by rubenfiszel** [comment]
Re: the latest Codex P1 (preserved `permissioned_as` resolving to a non-member superadmin) — this is not a regression from this PR and the suggested hardening would break legitimate behavior, so we're intentionally leaving it as-is.

**Not a new boundary.** `resolve_permissioned_as` stores the admin/`wm_deployer`-supplied value verbatim, and on `main` the bare-email spelling already escalates: `permissioned_as = "superadmin@windmill.dev"` → `get_email_from_permissioned_as` returns it via the no-prefix else branch → `fetch_authed_from_permissioned_as` else branch sets `is_admin: is_super_admin = true`. So the same privileged actors could already produce superadmin-owned execution before this PR — this change only adds an equivalent `u/<derived>` spelling for an identity they could already target. No new actor, capability, or privilege boundary.

**Why we can't "constrain to workspace members".** Membership validation is incompatible with what this PR supports:
- On the normal creation path, a non-member superadmin creating their *own* schedule/trigger stores `permissioned_as = u/<derived>` (they have no `usr` row). Rejecting non-members there would break the core scenario this PR enables.
- On the preserve path, a non-member-superadmin-owned schedule that is deployed or git-synced into another workspace preserves `u/<derived>`; membership validation would reject the deploy/sync of any non-member-owned item.

`preserve_permissioned_as` is by design a trusted admin/deployer capability to stamp arbitrary ownership (incl. non-members / superadmins); the only thing separating "hostile" from "legitimate" is caller authorization, and admins/deployers *are* the authorized set — which is the pre-existing trust model, unchanged here.

Net: pre-existing, net-neutral, and by-design. The added schedule test covers the intended self-owned non-member-superadmin path; the "hostile preserve" case is the same trusted-capability behavior that predates this PR.
